### PR TITLE
Issue/27 release automation

### PR DESCRIPTION
**Pull request by sanderr**
- added `ci/Jenkinsfile-pyp` to do the release. It sets the release date, adds a tag and opens a pull request for the version bump, similar to the pyp release pipeline for `inmanta-core` but without the `next` release.
- added Jenkins job https://jenkins.inmanta.com/job/releases/job/python-packages/job/sphinx-inmanta-pypi-release-automation-27/. Will rename to `sphinx-inmanta-pypi-release` and remove old release and dev release jobs once the PR is merged in.
- moved `Jenkinsfile` to `ci/Jenkinsfile-test`. Will update https://jenkins.inmanta.com/job/docs/job/inmanta-sphinx/ once the PR is merged in.

### CODE REVIEW

**Comment by sanderr** [comment]
@arnaudsjs @wouterdb make sure to read the description.

File: ci/Jenkinsfile-pyp [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+    parameters {
+        choice(
+            name: 'release_type',
+            choices: ['dev', 'stable'],
+            description: 'Indicate whether a dev or stable release should be built.'
+        )
+        choice(
+            name: 'version_increment',
+            choices: ['patch', 'minor', 'major'],
```
Comment (wouterdb) [inline ci/Jenkinsfile-pyp]: @arnaudsjs do we also need an option for no increment, when re-building a previously tagged release (e.g. pypi upload problems, ...)

File: ci/Jenkinsfile-pyp [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+            description: 'Indicates what kind of version bump should be done'
+        )
+        booleanParam(name: 'run_job', defaultValue: false, description: 'Allow for dry-runs')
```
Comment (wouterdb) [inline ci/Jenkinsfile-pyp]: This could do with a bit more explanation of what is done/not done when this box is ticked

File: ci/Jenkinsfile-pyp [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+                    usernamePassword(
+                        credentialsId: 'inmantaci_on_pypi',
+                        passwordVariable: 'TWINE_PASSWORD',
+                        usernameVariable: 'TWINE_USERNAME'
+                    ),
+                    string(credentialsId: 'package_gpg_passphrase', variable: 'PGP_PASS')
+                ]) {
+                    dir('inmanta-sphinx') {
+                        sh '''
+                            ${WORKSPACE}/env/bin/python3 -m twine upload dist/*.tar.gz
```
Comment (wouterdb) [inline ci/Jenkinsfile-pyp]: does this also sign? That used to be a separated gpg command?

**Review by wouterdb — COMMENTED** [review]
I'm not super knowledgeable about this code, I leave this to arnaud.

File: ci/Jenkinsfile-pyp (hunk 1 above)
Comment (arnaudsjs) [reply ci/Jenkinsfile-pyp]: Yes. That would be handy. Just in case the pipeline fails after the versionbump step.

File: ci/Jenkinsfile-pyp [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+                }
+            }
+            steps{
+                dir('inmanta-sphinx') {
+                    withCredentials([string(credentialsId: 'fff7ef7e-cb20-4fb2-a93b-c5139463c6bf', variable: 'GITHUB_TOKEN')]) {
+                        sh '''
+                            export GIT_ASKPASS="${WORKSPACE}/ASK_PASS"
+                            ${WORKSPACE}/env/bin/python3 -m pip install -U pip bumpversion
+
+                            git checkout -b issue/version-bump
```
Comment (arnaudsjs) [inline ci/Jenkinsfile-pyp:77]: We should make sure the correct branch is checked out. 
- For a minor or major release: master
- For a patch release: next

File: ci/Jenkinsfile-pyp [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+
+                            # bump version if major release
+                            if [ "${version_increment}" == "major" ]; then
+                                ${WORKSPACE}/env/bin/python3 -m bumpversion --commit ${version_increment} CHANGELOG.md
+                            fi
+
+                            # set release date
+                            release_date=$(date --utc +'%y-%m-%d')
+                            sed -i "1s/\\?/${release_date}/" changelog.md
+                            git commit -m "set release date: ${release_date}" changelog.md
```
Comment (arnaudsjs) [inline ci/Jenkinsfile-pyp]:
```suggestion
                            sed -i "1s/\\?/${release_date}/" CHANGELOG.md
                            git commit -m "set release date: ${release_date}" CHANGELOG.md
```

File: ci/Jenkinsfile-pyp [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+                            # Merge into next
+                            git checkout next
+                            git pull
+                            git merge -X theirs issue/version-bump
+                            git push origin next
+
+                            # Create a pull request to bump the version on master
+                            git checkout issue/version-bump
+                            ${WORKSPACE}/env/bin/python3 -m bumpversion "${version_increment}"
```
Comment (arnaudsjs) [inline ci/Jenkinsfile-pyp]:
```suggestion
                            ${WORKSPACE}/env/bin/python3 -m bumpversion minor
```
The version of master should be the next minor release compared to the latest stable release.

File: ci/Jenkinsfile-pyp [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+            }
+        }
+
+        stage("Build dist") {
+            when {
+                expression { "${run_job}" == "true" }
+                anyOf {
+                    expression { "${release_type}" == "stable" }
+                    expression { "${release_type}" == "dev" }
+                }
```
Comment (arnaudsjs) [inline ci/Jenkinsfile-pyp:171]: This condition can be removed. It's an exhaustive list of all possible values for release_type.

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: ci/Jenkinsfile-pyp [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+            }
+        }
+
+        stage("Publish to PyPi") {
+            when {
+                allOf {
+                    expression { "${run_job}" == "true" }
+                    expression { "${upload}" == "true" }
+                    expression { "${release_type}" == "stable" }
```
Comment (arnaudsjs) [inline ci/Jenkinsfile-pyp]: The previous build step allows building a dev release, but this step doesn't allow you to upload it to PyPi.

**Comment by arnaudsjs** [comment]
Also synchronize the latest branch with the next branch after a stable release is finished.

File: ci/Jenkinsfile-pyp (hunk 3 above)
Comment (sanderr) [reply ci/Jenkinsfile-pyp]: I don't know, but the [original job](https://jenkins.inmanta.com/job/releases/job/python-packages/job/sphinx-inmanta-pypi-release/) didn't contain any other steps.

File: ci/Jenkinsfile-pyp (hunk 7 above)
Comment (sanderr) [reply ci/Jenkinsfile-pyp:171]: I am aware. I added it as a guard against future expansion of `release_type` so we can be sure that the `else` branch is `dev`. Should I remove it?

Comment (arnaudsjs) [reply ci/Jenkinsfile-pyp:171]: Ok. You can keep it like this.

File: ci/Jenkinsfile-pyp (hunk 3 above)
Comment (wouterdb) [reply ci/Jenkinsfile-pyp]: see here: https://github.com/inmanta/irt/blob/master/ci/inmanta-core/Jenkinsfile-pyp#L224

**Review by wouterdb — APPROVED** [review]

File: ci/Jenkinsfile-pyp [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+        stage("Sync latest branch with next"){
+            when{
+                allOf {
+                    expression { "${run_job}" == "true" }
+                    expression { "${release_type}" == "stable" }
+                }
+            }
+
+            steps{
+                dir('inmanta') {
```
Comment (arnaudsjs) [inline ci/Jenkinsfile-pyp]:
```suggestion
                dir('inmanta-sphinx') {
```

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: ci/Jenkinsfile-pyp (hunk 3 above)
Comment (sanderr) [reply ci/Jenkinsfile-pyp]: @arnaudsjs Should I add a step like this? As mentioned above it isn't present in the current inmanta-sphinx release job but as Wouter shows it is present in the core release.

Comment (arnaudsjs) [reply ci/Jenkinsfile-pyp]: Yes. Signing packages is always good practice. I assume this step was forgotten during the initial setup of the release job.

File: ci/Jenkinsfile-pyp [hunk 10]
```diff
@@ -199,6 +202,8 @@ pipeline {
                 ]) {
                     dir('inmanta-sphinx') {
                         sh '''
+                            ${WORKSPACE}/env/bin/python3 -m pip install -U twine
+                            echo "${PGP_PASS}" | gpg --pinentry-mode loopback -u code@inmanta.com --passphrase-fd 0 --detach-sign --batch --yes -a dist/*.tar.gz
```
Comment (arnaudsjs) [inline ci/Jenkinsfile-pyp:206]: You should also upload the signature to PyPi. This is the file ending in `.asc`

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: ci/Jenkinsfile-pyp [hunk 11]
```diff
@@ -204,7 +204,7 @@ pipeline {
                         sh '''
                             ${WORKSPACE}/env/bin/python3 -m pip install -U twine
                             echo "${PGP_PASS}" | gpg --pinentry-mode loopback -u code@inmanta.com --passphrase-fd 0 --detach-sign --batch --yes -a dist/*.tar.gz
-                            ${WORKSPACE}/env/bin/python3 -m twine upload dist/*.tar.gz
+                            ${WORKSPACE}/env/bin/python3 -m twine upload dist/*.tar.{gz,asc}
```
Comment (arnaudsjs) [inline ci/Jenkinsfile-pyp]: The asc file is called `<name>.tar.gz.asc` and not `<name>.tar.asc`

File: ci/Jenkinsfile-pyp [hunk 12]
```diff
@@ -238,7 +238,8 @@ pipeline {
 
     post {
         success{
-            archiveArtifacts artifacts: 'inmanta-sphinx/dist/*.tar.gz', allowEmptyArchive: true
+            archiveArtifacts artifacts: 'inmanta-sphinx/dist/*.tar.gz, inmanta-sphinx/dist/*.tar.asc',
```
Comment (arnaudsjs) [inline ci/Jenkinsfile-pyp]: Same here.

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

**Review by arnaudsjs — APPROVED** [review]

**Pull request merged**